# Introdução a Limpeza de Dados 

A limpeza dos dados é um passo muito importante para que possamos deixar tudo pronto para a visualização dos dados.

Só a limpeza de dados consome de 60% a 80% do trabalho de um Cientista de Dados.

Nos deparamos com diversos erros em tabelas, dentre eles temos, nome de colunas erradas, muita informação sem utilidade, formato errado dos dados e outros.

Por isso a limpeza de dados é um papel fundamental e árduo, e que muitos não querem fazer... Mas você, que é um(a) guerreiro(a), está disposto(a) a fazer esse trabalho duro.

Então vamos lá, que eu vou te ajudar a tornar esse trabalho um pouco mais fácil.

Hoje teremos alguns exemplos de como podemos limpar nossos dados para melhor visualização dos mesmos.

## Recapitulação

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.df = pd.read_csv('world_population.csv')
df.head()

Rank CCA3         Country           Capital Continent  2022 Population  \
0    36  AFG     Afghanistan             Kabul      Asia         41128771   
1   138  ALB         Albania            Tirana    Europe          2842321   
2    34  DZA         Algeria           Algiers    Africa         44903225   
3   213  ASM  American Samoa         Pago Pago   Oceania            44273   
4   203  AND         Andorra  Andorra la Vella    Europe            79824   

   2020 Population  2015 Population  2010 Population  2000 Population  \
0         38972230         33753499         28189672         19542982   
1          2866849          2882481          2913399          3182021   
2         43451666         39543154         35856344         30774621   
3            46189            51368            54849            58230   
4            77700            71746            71519            66097   

   1990 Population  1980 Population  1970 Population  Area (km²)  \
0         10694796         12486631         10752971      652230   
1          3295066          2941651          2324731       28748   
2         25518074         18739378         13795915     2381741   
3            47818            32886            27075         199   
4            53569            35611            19860         468   

   Density (per km²)  Growth Rate  World Population Percentage  
0            63.0587       1.0257                         0.52  
1            98.8702       0.9957                         0.04  
2            18.8531       1.0164                         0.56  
3           222.4774       0.9831                         0.00  
4           170.5641       1.0100                         0.00

Vamos tentar fazer algumas recuperações de dados **sem** utilizar **loc e iloc**

Que tal iniciar com os dados referentes ao Brasil?

In [ ]:
Brasil = df[df["Country"] == "Brazil"]
Brasil

Rank CCA3 Country   Capital      Continent  2022 Population  \
27     7  BRA  Brazil  Brasilia  South America        215313498   

    2020 Population  2015 Population  2010 Population  2000 Population  \
27        213196304        205188205        196353492        175873720   

    1990 Population  1980 Population  1970 Population  Area (km²)  \
27        150706446        122288383         96369875     8515767   

    Density (per km²)  Growth Rate  World Population Percentage  
27            25.2841       1.0046                          2.7

Agora vamos tentar relembrar do método **sum**, que tal descobrirmos qual a **população total** de 2022? 

In [ ]:
df["2022 Population"].sum()

7973413042

E que tal a média das Áreas dos países? Lembre-se que o método **mean** pode nos auxiliar com isso!

In [ ]:
df["Area (km²)"].mean()

581449.3846153846

Agora vamos lembrar de como obter todas as **linhas** que encontram uma **determinada condição**, que tal tentarmos obter todos os países da América do Sul?

In [ ]:
AdS = df[df["Continent"] == "South America"]
AdS

Rank CCA3           Country       Capital      Continent  \
8      33  ARG         Argentina  Buenos Aires  South America   
24     80  BOL           Bolivia         Sucre  South America   
27      7  BRA            Brazil      Brasilia  South America   
40     65  CHL             Chile      Santiago  South America   
42     28  COL          Colombia        Bogota  South America   
56     67  ECU           Ecuador         Quito  South America   
64    231  FLK  Falkland Islands       Stanley  South America   
69    184  GUF     French Guiana       Cayenne  South America   
86    164  GUY            Guyana    Georgetown  South America   
161   109  PRY          Paraguay      Asunción  South America   
162    44  PER              Peru          Lima  South America   
199   170  SUR          Suriname    Paramaribo  South America   
223   133  URY           Uruguay    Montevideo  South America   
227    51  VEN         Venezuela       Caracas  South America   

     2022 Population  2020 Population  2015 Population  2010 Population  \
8           45510318         45036032         43257065         41100123   
24          12224110         11936162         11090085         10223270   
27         215313498        213196304        205188205        196353492   
40          19603733         19300315         17870124         17004162   
42          51874024         50930662         47119728         44816108   
56          18001000         17588595         16195902         14989585   
64              3780             3747             3408             3187   
69            304557           290969           257026           228453   
86            808726           797202           755031           747932   
161          6780744          6618695          6177950          5768613   
162         34049588         33304756         30711863         29229572   
199           618040           607065           575475           546080   
223          3422794          3429086          3402818          3352651   
227         28301696         28490453         30529716         28715022   

     2000 Population  1990 Population  1980 Population  1970 Population  \
8           37070774         32637657         28024803         23842803   
24           8592656          7096194          5736088          4585693   
27         175873720        150706446        122288383         96369875   
40          15351799         13342868         11469828          9820481   
42          39215135         32601393         26176195         20905254   
56          12626507         10449837          8135845          6172215   
64              3080             2332             2240             2274   
69            164351           113931            66825            46484   
86            759051           747116           778176           705261   
161          5123819          4059195          3078912          2408787   
162         26654439         22109099         17492406         13562371   
199           478998           412756           375112           379918   
223          3292224          3117012          2953750          2790265   
227         24427729         19750579         15210443         11355475   

     Area (km²)  Density (per km²)  Growth Rate  World Population Percentage  
8       2780400            16.3683       1.0052                         0.57  
24      1098581            11.1272       1.0120                         0.15  
27      8515767            25.2841       1.0046                         2.70  
40       756102            25.9274       1.0057                         0.25  
42      1141748            45.4339       1.0069                         0.65  
56       276841            65.0229       1.0114                         0.23  
64        12173             0.3105       1.0043                         0.00  
69        83534             3.6459       1.0239                         0.00  
86       214969             3.7621       1.0052                   

Agora que conseguimos obter esses países, vamos utilizar esse novo dataset nas nossas próximas recuperações de dados.

Vamos começar procurando os países com as **5 maiores populações** da América do Sul, não esqueça que o método **nlargest** pode nos auxiliar nesse quesito!

In [ ]:
AdS.nlargest(5, ["2022 Population"])

Rank CCA3    Country       Capital      Continent  2022 Population  \
27      7  BRA     Brazil      Brasilia  South America        215313498   
42     28  COL   Colombia        Bogota  South America         51874024   
8      33  ARG  Argentina  Buenos Aires  South America         45510318   
162    44  PER       Peru          Lima  South America         34049588   
227    51  VEN  Venezuela       Caracas  South America         28301696   

     2020 Population  2015 Population  2010 Population  2000 Population  \
27         213196304        205188205        196353492        175873720   
42          50930662         47119728         44816108         39215135   
8           45036032         43257065         41100123         37070774   
162         33304756         30711863         29229572         26654439   
227         28490453         30529716         28715022         24427729   

     1990 Population  1980 Population  1970 Population  Area (km²)  \
27         150706446        122288383         96369875     8515767   
42          32601393         26176195         20905254     1141748   
8           32637657         28024803         23842803     2780400   
162         22109099         17492406         13562371     1285216   
227         19750579         15210443         11355475      916445   

     Density (per km²)  Growth Rate  World Population Percentage  
27             25.2841       1.0046                         2.70  
42             45.4339       1.0069                         0.65  
8              16.3683       1.0052                         0.57  
162            26.4933       1.0099                         0.43  
227            30.8820       1.0036                         0.35

### Exercícios

In [ ]:
# Obtenha os 5 maiores países do dataset original (o dataset "df")

In [ ]:
# Obtenha todos os países da Europa e imprima apenas os 5 primeiros países, não esqueça que o método .head() pode nos ajudar!

In [ ]:
# Agora faça a média da população de 2022 da Europa

## Lidando com Strings

Uma das formas de limpar seu dataset é trabalhando com strings, então vamos ver algumas operações com o dataset pessoas.csv.

Antes de tudo precisamos carregar o dataset.

In [1]:
import pandas as pd

df = pd.read_csv('pessoas.csv')
df

nome                      em                   endereco  \
0  Ana souza roCHA      anasouza@gmail.com  Nossa Senhora do Carmo 68   
1  Cristina BARROS     crisbarros@gmal.com                        NaN   
2       Joao Alves   alvesjoao@hotmail.com                        NaN   
3     mario santos      m.santos@gmail.com      Rua Joao Gilberto 900   
4              NaN   cferreira@hotmail.com      Rua XV de novembro 89   
5   MARIA DA SILVA       dasilva@gmail.com      Rua das Oliveiras 400   

          cidade  
0       Alegrete  
1       Alegrete  
2       Alegrete  
3     Uruguaiana  
4     Uruguaiana  
5     Uruguaiana

Analisando um pouco o dataset percebemos bastante erros que precisamos arrumar.

Primeiro vamos para a coluna nome. Aqui podemos fazer algumas alterações como você pode ver abaixo.

In [ ]:
#Colocamos todos os nomes em caixa baixa ou lower case
df.nome = df.nome.str.lower()
df

In [ ]:
#Podemos também dar um capitalize que seria colocar a primeira letra em maiusculo
df.nome = df.nome.str.capitalize()
df

In [ ]:
#Ou transformar em titulo onde todas as primeiras letras serão maiusculas
df.nome = df.nome.str.title()
df

Conseguimos com a função replace, trocar strings do nosso dataset.

Por exemplo:

In [ ]:
df.em = df.em.str.replace('gmal', 'gmail')
df

Uma função muito útil do str é o contains que retorna true ou false se a string contém o que você colocar dentro dos parenteses.

Vamos verificar se na nossa coluna cidade existe espaços em brancos:

In [ ]:
df.cidade.str.contains(' ').value_counts()

Temos 5 cidades com espaço em branco podemos arrumar isso com o replace:

In [ ]:
#Remova os espaços em branco no inicio do nome das cidades. 
df.cidade = df.cidade.str.replace('\s+','')

#verifica se ainda existe espaços em branco
df.cidade.str.contains(' ')

Vamos ver um pouco de regex, que seria uma expressão para achar strings com o padrão que você quiser.

Por exemplo essa expressão [AEIOU] vai nos retornar todas as strings que contem essas letras, tome cuidado que os regex são case sensitive.

Vamos ver alguns exemplos:

In [3]:
#Quantas pessoas possuem nome próprio começando com consoante e quantas com vogal?
s = df.nome.str.count('^[^AEIOU]')
s.value_counts()

0    0.0
1    1.0
2    1.0
3    1.0
4    NaN
5    1.0
Name: nome, dtype: float64

In [ ]:
#Quais os primeiros nomes das pessoas ?

df.nome.str.extract('([A-Za-z]+)', expand=False)

In [ ]:
#Quais os primeiros nomes de usuário de email ?
import re

df.em.str.extract('([A-Za-z0-9._%+-]+@)', expand=False)

Aqui foram alguns exemplos, mas você pode ver mais do regex na <a href='https://docs.python.org/3/howto/regex.html#regex-howto'>documentação do python</a>.

## Dataset sobre livros e suas informações.


Ele nos mostra uma serie de informações sobre alguns livros publicados, como data de publicação, local de publicação, entre outros.

Vamos trabalhar um pouco com ele, pois é um dataset muito bom para termos uma ideia melhor de como fazer a limpeza de um dataset.

In [ ]:
df = pd.read_csv('BL-Flickr-Images-Book.csv')
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

Como você pode ter visto nossa tabela tem vários dados faltantes e isso na verdade é bem comum de acontecer, então saber lidar com os dados faltantes (ou NaN) é fundamental para um Cientista de Dados.

### Drop de colunas com valores faltantes.

Para iniciarmos vamos fazer uso de uma função muito boa do pandas que é o .dropna, com ela conseguimos lidar bem com os dados faltantes.

O .dropna sozinho tira todas as linhas que tenham alguma coluna com NaN, mas não exatamente o que queremos aqui, adicionando alguns parâmetros conseguimos chegar em uma solução boa para o nosso problema.

Primeiro o how='all', estamos falando que queremos tirar somente aqueles que estejam com todos os dados faltantes.

Segundo axis=1, aqui estamos olhando para as colunas e não para as linhas, estamos avisando a função disso.

E por último o inplace=true, estamos pedindo para a função fazer as modificações na própria tabela, pois por definição ela cria uma nova tabela com as modificações feitas.

In [ ]:
df.dropna(how='all',axis=1,inplace=True)
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors Former owner Issuance type  \
0                               FORBES, Walter.          NaN   monographic   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness          NaN   monographic   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness          NaN   monographic   
3                   Appleyard, Ernest Silvanus.          NaN   monographic   
4                           BROOME, John Henry.          NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

Lidar com dados faltantes é muito importante, e você precisa saber o que fazer com eles, terá vezes que somente tirar eles dos datasets não será a melhor opção.

Dito isso você pode pesquisar mais sobre o dropna e outras funções na <a href='https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html'>documentação do pandas</a>.

### Drop de colunas em especifico

Continuando com a limpeza nessa tabela, temos muitas informações ainda e colunas com muitos dados faltando, nesse caso específico as colunas que iremos usar são as que não estão faltando dados.

Então podemos remover essas colunas que não queremos.

Aqui utilizamos a função drop do pandas, podemos passar uma lista completa do que queremos remover da nossa lista.

Aqui de novo temos dois parâmetros já usados, o inplace e o axis.

In [ ]:
to_drop = ['Edition Statement',
           'Former owner',
           'Contributors',
           'Issuance type',
           'Shelfmarks']

df.drop(to_drop, inplace=True, axis=1)

In [ ]:
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                         Flickr URL  
0      A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
1  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
2  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
3  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...  
4  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...

### Troca de coluna índice

Como expliquei o pandas cria uma coluna de identificação por padrão, para facilitar nossa visualização podemos colocar uma nova coluna como identificação, porém essa coluna precisa ser única, ou seja, ela precisa ser diferente para cada linha da nossa tabela.

No codigo abaixo utilizamos de uma função do pandas onde ele nos retorna se uma coluna da nossa tabela é única ou não. Bem simples não?

In [ ]:
df['Identifier'].is_unique

True

Como vimos que nossa coluna Identifier é unica podemos colocar ela como nossa coluna de identificação e facilitar nosso trabalho no restante da limpeza.

In [ ]:
df = df.set_index('Identifier')
#df.set_index('Identifier', inplace=True)
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

### Normalização de colunas

Agora pra começar nossa limpeza de verdade vamos analisar algumas colunas e ver se elas estão de acordo com o que a gente quer.

Aqui utilizaremos uma função do pandas para saber qual o tipo dos dados da nossa coluna, nesse caso vamos ver a data de publicação, queremos então que ela seja de alguma forma do tipo inteiro ou float.

Vamos ver qual o tipo da nossa data de publicação:

In [ ]:
df['Date of Publication'].dtype

dtype('O')

Tipo 'O', hmmmmm.

Acredito que alguma coisa esteja errada, vamos analisar então o que temos nessa coluna, já que o tipo dela está errado, podemos encontrar coisas que também não queremos.

Vamos utilizar uma função muito boa do pandas que é o .loc, com essa função podemos resgatar nossos dados de diversas formas, nesse caso estamos pegando todos os identifier do 1905 pra frente e resgatando a coluna Date of Publication tambem e só mostrando as 10 primeiras.
Por isso foi bom mudar a nossa coluna identificação.

In [ ]:
df.loc[1905:, 'Date of Publication'].head(10)

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
2957           1873
3017           1866
3131           1899
4598           1814
4884           1820
Name: Date of Publication, dtype: object

Olha só, parece que nossa coluna não tá muito legal.

Vamos arrumar isso com muita facilidade utilizando uma função regex.

"O que é uma regex?" você pode estar se perguntando. Uma regex é uma string de padronização, quando por exemplo queremos que todos os email tenham o '@' alguma coisa ou todas as datas sigam DD/MM/AAAA.

Entendeu?

In [ ]:
r = r'^(\d{4})'
extr = df['Date of Publication'].str.extract(r, expand=False)
extr.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

O que aconteceu aqui em cima foi que tiramos todos os 4 dígitos que ficam juntos, que é nossa data de publicação.

Agora além de colocar nossa data certinha, podemos arrumar o tipo dela, lembra que ela estava do tipo 'O', então agora vamos colocar para tipo numérico onde podemos trabalhar de forma mais fácil.

:)

In [ ]:
df['Date of Publication'] = pd.to_numeric(extr)
df['Date of Publication'].dtype

dtype('float64')

In [ ]:
df['Date of Publication'].isnull().sum()

971

Com o tipo numérico podemos fazer como foi mostrado a cima, somar todos os valores que são nulos, o que nos dá 971 numeros nulos na nossa tabela. 😧

Calma lá, vamos ver o que esses 971 influenciam na nossa tabela para saber se teremos que cuidar deles ou não.

In [ ]:
df['Date of Publication'].isnull().sum() / len(df)

0.021841438397490046

Então esses 971 representam 11% da nossa tabela.
 
É uma porcentagem significante, mas ainda assim pode não afetar nosso trabalho.

Bora pra proxima coluna, dessa vez vamos dar uma olhada na nossa coluna Place of Publication, que seria uma forte candidata para nossas futuras visualizações como a data de publicação.

In [ ]:
df['Place of Publication'].head(10)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
Name: Place of Publication, dtype: object

Não parece estar muito bonita nossa coluna, vamos fazer algumas alterações.

Aqui utilizaremos uma abordagem diferente, vamos utilizar o .contains e uma função do numpy np.where que basicamente ele faz alguma coisa se for verdade e outra coisa se for falso.

Vamos ver na prática:

In [ ]:
pub = df['Place of Publication']
london = pub.str.contains('London')
london[:5]

Identifier
206    True
216    True
218    True
472    True
480    True
Name: Place of Publication, dtype: bool

Temos agora um lista com todas as posições onde tem London no nome.

Vamos fazer a mesma coisa com Oxford

In [ ]:
oxford = pub.str.contains('Oxford')

Agora vamos usar o np.where. A leitura do código é assim, onde na lista nossa condição for verdadeira, vamos colocar London. Quando for falsa, a gente coloca outro np.where pra executarmos uma outra verificação. Então, quando nossa primeira condição for falsa, quero que entre em outro where que vai colocar Oxford para verdadeiros e trocar '-' por ' ' para os itens falsos.

Ficou confuso? Basicamente criamos duas listas onde elas têm a localização de onde tem London e Oxford na nossa lista original, então eu só pedi pra trocar o nome todo por esses dois.

In [ ]:
#np.where function is called in a nested structure

df['Place of Publication'] = np.where(london, 'London',
                                      np.where(oxford, 'Oxford',
                                               pub.str.replace('-', ' ')))

df['Place of Publication'].head()

Identifier
206    London
216    London
218    London
472    London
480    London
Name: Place of Publication, dtype: object

In [ ]:
df.head()

Place of Publication  Date of Publication              Publisher  \
Identifier                                                                    
206                      London               1879.0       S. Tinsley & Co.   
216                      London               1868.0           Virtue & Co.   
218                      London               1869.0  Bradbury, Evans & Co.   
472                      London               1851.0          James Darling   
480                      London               1857.0   Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

Por fim nossa tabela ficou um pouco mais organizada ao queremos fazer, o que quer que seja isso.

## Dataset exemplo sobre olimpíadas

O dataset pode ser encontrado por esse link -> <a href="https://drive.google.com/file/d/1eX6usv2GmThnhW1KuBst8vPkMSmNVMdr/view?usp=sharing">dataset olimpiadas</a>

Ele nos traz algumas informações sobre olimpíadas, medalhas e o número de jogos jogados.

Esse dataset é interessante por nos trazer alguns erros que não vimos antes, vamos ver como arrumar eles.

In [ ]:
olympics_df = pd.read_csv('olimpics.csv')
olympics_df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  ? Summer  01 !  02 !  03 !  Total  ? Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  ? Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

Essa tabela está bem dificil de entender alguma coisa.

De primeira vemos que ele leu errado na hora de carregar os dados e colocou outra linha de índicies para as colunas.

Vamos começar por aí.

In [ ]:
olympics_df = pd.read_csv('olimpics.csv', header=1)
olympics_df.head()

Unnamed: 0  ? Summer  01 !  02 !  03 !  Total  ? Winter  \
0        Afghanistan (AFG)        13     0     0     2      2         0   
1            Algeria (ALG)        12     5     2     8     15         3   
2          Argentina (ARG)        23    18    24    28     70        18   
3            Armenia (ARM)         5     1     2     9     12         6   
4  Australasia (ANZ) [ANZ]         2     3     4     5     12         0   

   01 !.1  02 !.1  03 !.1  Total.1  ? Games  01 !.2  02 !.2  03 !.2  \
0       0       0       0        0       13       0       0       2   
1       0       0       0        0       15       5       2       8   
2       0       0       0        0       41      18      24      28   
3       0       0       0        0       11       1       2       9   
4       0       0       0        0        2       3       4       5   

   Combined total  
0               2  
1              15  
2              70  
3              12  
4              12

Aqui nós só falamos pra que na hora de carregar os dados não considerar a primeira linha e que os índices das nossas colunas estão na próxima linha.

Mas claro que não da pra parar por aí, dá uma olhada nos nomes das nossas colunas, nessa parte vai um pouco de achismo de qual nome colocar aqui.

Qual nome você daria pra cada coluna?

In [ ]:
new_names =  {'Unnamed: 0': 'Country',
              '? Summer': 'Summer Olympics',
              '01 !': 'Gold',
              '02 !': 'Silver',
              '03 !': 'Bronze',
              '? Winter': 'Winter Olympics',
              '01 !.1': 'Gold.1',
              '02 !.1': 'Silver.1',
              '03 !.1': 'Bronze.1',
              '? Games': '# Games',
              '01 !.2': 'Gold.2',
              '02 !.2': 'Silver.2',
              '03 !.2': 'Bronze.2'}

olympics_df.rename(columns=new_names, inplace=True)
olympics_df.head()

Country  Summer Olympics  Gold  Silver  Bronze  Total  \
0        Afghanistan (AFG)               13     0       0       2      2   
1            Algeria (ALG)               12     5       2       8     15   
2          Argentina (ARG)               23    18      24      28     70   
3            Armenia (ARM)                5     1       2       9     12   
4  Australasia (ANZ) [ANZ]                2     3       4       5     12   

   Winter Olympics  Gold.1  Silver.1  Bronze.1  Total.1  # Games  Gold.2  \
0                0       0         0         0        0       13       0   
1                3       0         0         0        0       15       5   
2               18       0         0         0        0       41      18   
3                6       0         0         0        0       11       1   
4                0       0         0         0        0        2       3   

   Silver.2  Bronze.2  Combined total  
0         0         2               2  
1         2         8              15  
2        24        28              70  
3         2         9              12  
4         4         5              12

Veja que a gente passou um dicionário pra renomear as nossas colunas! Lembra dos dicionários? São aquela estrutura de dados que temos uma lista de itens em que cada dado tem seu nome. Lembrou? Então vamos entender como aconteceu a renomeação ali em cima.

Para fazer a renomeação das nossas colunas, nós criamos uma variável pra carregar nosso dicionário. Colocamos os nomes antigos como as **chaves** do dicionário e os nomes novos como os **valores** do dict. Quando colocamos nosso dicionário no método rename(), o pandas vai interpretar nosso dicionário e substituir os nomes das colunas pra gente.

Aí tá uma tabela mais legal de se mexer, não acha?

## Exercício do dataset da população

Vamos ao nosso último exemplo.

Pegaremos o dataset que estávamos trabalhando nas últimas semanas.

Lembra de como carregar ele pro notebook?

In [ ]:
#Escreva aqui o codigo para carregar o notebook

Rank CCA3 Country/Territory           Capital Continent  2022 Population  \
0    36  AFG       Afghanistan             Kabul      Asia         41128771   
1   138  ALB           Albania            Tirana    Europe          2842321   
2    34  DZA           Algeria           Algiers    Africa         44903225   
3   213  ASM    American Samoa         Pago Pago   Oceania            44273   
4   203  AND           Andorra  Andorra la Vella    Europe            79824   

   2020 Population  2015 Population  2010 Population  2000 Population  \
0         38972230         33753499         28189672         19542982   
1          2866849          2882481          2913399          3182021   
2         43451666         39543154         35856344         30774621   
3            46189            51368            54849            58230   
4            77700            71746            71519            66097   

   1990 Population  1980 Population  1970 Population  Area (km²)  \
0         10694796         12486631         10752971      652230   
1          3295066          2941651          2324731       28748   
2         25518074         18739378         13795915     2381741   
3            47818            32886            27075         199   
4            53569            35611            19860         468   

   Density (per km²)  Growth Rate  World Population Percentage  
0            63.0587       1.0257                         0.52  
1            98.8702       0.9957                         0.04  
2            18.8531       1.0164                         0.56  
3           222.4774       0.9831                         0.00  
4           170.5641       1.0100                         0.00

Vamos fazer poucas alterações no nosso dataset, pq ele está bem completo e sem muitos erros (uma coisa fora do comum).

Vamos começar reduzindo nossos dados, pois nesse caso queremos apenas dados mais recentes.

Fique com apenas a coluna de 2022 e tire todas as outras.

In [ ]:
#Escreva aqui o codigo
#Dica, vimos algo assim no dataset dos livros

In [ ]:
#Mostre os 5 primeiros da tabela

Rank CCA3 Country/Territory           Capital Continent  2022 Population  \
0    36  AFG       Afghanistan             Kabul      Asia         41128771   
1   138  ALB           Albania            Tirana    Europe          2842321   
2    34  DZA           Algeria           Algiers    Africa         44903225   
3   213  ASM    American Samoa         Pago Pago   Oceania            44273   
4   203  AND           Andorra  Andorra la Vella    Europe            79824   

   Area (km²)  Density (per km²)  Growth Rate  World Population Percentage  
0      652230            63.0587       1.0257                         0.52  
1       28748            98.8702       0.9957                         0.04  
2     2381741            18.8531       1.0164                         0.56  
3         199           222.4774       0.9831                         0.00  
4         468           170.5641       1.0100                         0.00

Mostre que a coluna que você escolheu é única.

Qual coluna você acha que seria uma boa coluna de identificação?

In [ ]:
#Escreva seu codigo aqui

True

Agora que vimos que nossa coluna é única, podemos transformar ela em nossa coluna de indentificação.

Lembra de como que faz?

In [ ]:
#Escreva aqui

Rank Country/Territory           Capital Continent  2022 Population  \
CCA3                                                                        
AFG     36       Afghanistan             Kabul      Asia         41128771   
ALB    138           Albania            Tirana    Europe          2842321   
DZA     34           Algeria           Algiers    Africa         44903225   
ASM    213    American Samoa         Pago Pago   Oceania            44273   
AND    203           Andorra  Andorra la Vella    Europe            79824   

      Area (km²)  Density (per km²)  Growth Rate  World Population Percentage  
CCA3                                                                           
AFG       652230            63.0587       1.0257                         0.52  
ALB        28748            98.8702       0.9957                         0.04  
DZA      2381741            18.8531       1.0164                         0.56  
ASM          199           222.4774       0.9831                         0.00  
AND          468           170.5641       1.0100                         0.00

Podemos renomear algumas colunas para que nossa visualização fique melhor.

Vamos renomear aquelas em inglês e colocar em português.

In [ ]:
#Coloque aqui o codigo
#Dica: fizemos isso no dataset das olimpiadas

In [ ]:
#Mostre os 5 primeiros

Rank            País           Capital Continente  População em 2022  \
CCA3                                                                         
AFG     36     Afghanistan             Kabul       Asia           41128771   
ALB    138         Albania            Tirana     Europe            2842321   
DZA     34         Algeria           Algiers     Africa           44903225   
ASM    213  American Samoa         Pago Pago    Oceania              44273   
AND    203         Andorra  Andorra la Vella     Europe              79824   

      Area (km²)  Densidade (km²)  Crescimento Populacional  \
CCA3                                                          
AFG       652230          63.0587                    1.0257   
ALB        28748          98.8702                    0.9957   
DZA      2381741          18.8531                    1.0164   
ASM          199         222.4774                    0.9831   
AND          468         170.5641                    1.0100   

      Porcentagem com Relação ao Mundo  
CCA3                                    
AFG                               0.52  
ALB                               0.04  
DZA                               0.56  
ASM                               0.00  
AND                               0.00

Aqui temos uma coisinha nova então vou deixar para vocês.

Estamos trocando o nome dos valores que aparecem na coluna Continente.

In [ ]:
df.replace({'Continente':{'Europe' : 'Europa', 'North America' : 'América do Norte', 'South America' : 'América do Sul'}})

Rank               País           Capital Continente  População em 2022  \
CCA3                                                                            
AFG     36        Afghanistan             Kabul       Asia           41128771   
ALB    138            Albania            Tirana     Europa            2842321   
DZA     34            Algeria           Algiers     Africa           44903225   
ASM    213     American Samoa         Pago Pago    Oceania              44273   
AND    203            Andorra  Andorra la Vella     Europa              79824   
...    ...                ...               ...        ...                ...   
WLF    226  Wallis and Futuna          Mata-Utu    Oceania              11572   
ESH    172     Western Sahara          El Aaiún     Africa             575986   
YEM     46              Yemen             Sanaa       Asia           33696614   
ZMB     63             Zambia            Lusaka     Africa           20017675   
ZWE     74           Zimbabwe            Harare     Africa           16320537   

      Area (km²)  Densidade (km²)  Crescimento Populacional  \
CCA3                                                          
AFG       652230          63.0587                    1.0257   
ALB        28748          98.8702                    0.9957   
DZA      2381741          18.8531                    1.0164   
ASM          199         222.4774                    0.9831   
AND          468         170.5641                    1.0100   
...          ...              ...                       ...   
WLF          142          81.4930                    0.9953   
ESH       266000           2.1654                    1.0184   
YEM       527968          63.8232                    1.0217   
ZMB       752612          26.5976                    1.0280   
ZWE       390757          41.7665                    1.0204   

      Porcentagem com Relação ao Mundo  
CCA3                                    
AFG                               0.52  
ALB                               0.04  
DZA                               0.56  
ASM                               0.00  
AND                               0.00  
...                                ...  
WLF                               0.00  
ESH                               0.01  
YEM                               0.42  
ZMB                               0.25  
ZWE                               0.20  

[234 rows x 9 columns]